envisage/plugins/ipython_kernel/heartbeat.py

```diff
@@ -0,0 +1,43 @@
+"""
+A simple ping-pong style heartbeat that runs in a thread.
+
+Modified from upstream to enable the heartbeat thread to be shut down cleanly.
+"""
+# We're currently using ipykernel v4.10.1. In that version, there's no way to
```

Comment:
I don't think this requirement encoded anywhere other than implicitly what we get from EDM? There should be an optional setup.py requirement added for ipython support for folks getting things from `pip`.

Reply:
Ah, good point. I need to test whether this code actually works with all the latest versions from PyPI. It might, or it might not. If it does, great. If not, but there are easy fixes, I'll make those fixes. If not, and there are no easy fixes, we'll have to open issues and add version constraints in setup.py.

This *particular* hack should be forward-compatible: it'll work just as well with ipykernel 5.x, because that version already catches and deals with the exception, so our `except` block here won't get exercised.

Hmm; that does mean that the socket won't get deleted, though. I may still have to deal with that. Will check.